src/visitors/defuse_analyze_visitor.cpp

```diff
@@ -127,23 +130,27 @@ DUState DUInstance::sub_block_eval() const {
  *    block encountered, this means every block has either "D" or "CD". In
  *    this case we can say that entire block effectively has "D".
  */
-DUState DUInstance::conditional_block_eval() const {
+DUState DUInstance::conditional_block_eval(
+    DUVariableType variable_type = DUVariableType::Global) const {
     DUState result = DUState::NONE;
     bool block_with_none = false;
 
     for (const auto& chain: children) {
-        auto child_state = chain.eval();
-        if (child_state == DUState::U) {
+        auto child_state = chain.eval(variable_type);
+        if ((variable_type == DUVariableType::Global && child_state == DUState::U) ||
+            (variable_type == DUVariableType::Local && child_state == DUState::LU)) {
             result = child_state;
             break;
         }
         if (child_state == DUState::NONE) {
             block_with_none = true;
         }
-        if (child_state == DUState::D || child_state == DUState::CD) {
+        if ((variable_type == DUVariableType::Global && child_state == DUState::D) ||
+            (variable_type == DUVariableType::Local && child_state == DUState::LD) ||
+            child_state == DUState::CD) {
             result = DUState::CD;
             if (chain.state == DUState::ELSE && !block_with_none) {
-                result = DUState::D;
+                result = child_state;
                 break;
             }
         }
```

Comment:
unit tests would be great!

Reply:
Added a couple of them that cover some important cases I believe